### DISCUSSION TITLE
Adding completions to a formula when the upstream does not ship with them

### DISCUSSION THREAD

**Discussion by snim2** [Homebrew/discussions#3709] [Writing Formulae/Casks] [answered]
The [`adr-tools`](https://github.com/npryce/adr-tools) package has a [Homebrew formula](https://formulae.brew.sh/formula/adr-tools) and it ships with [Bash completions](https://github.com/npryce/adr-tools/tree/master/autocomplete) but not completions for `zsh` or other shells. I'd like to add `zsh` completions, and I have a script for this that works. 

Unfortunately `adr-tools` is essentially unmaintained, and I think the chances of getting a new completions file merged in are very slim. The Homebrew formula seems to get a decent number of installs for a shell utility, so writing a new formula for a fork seems unhelpful.

Would it be sensible to just write the completions script out directly from the existing formula, like this:

```ruby
zsh_completions = buildpath/"autocomplete/_adr"
zsh_completions.write <<~EOS
  ...
EOS
zsh_completion.install "autocomplete/_adr"
```

Or to ship the completions script separately, and edit the formula to install it from GitHub? Or is there a better way forward?

For context, I've tried adding:

```ruby
generate_completions_from_executable(bin/"adr", "completion", shells: [:zsh])
```

which fails without any output:

```shell
> brew reinstall --verbose --build-from-source adr-tools
...
Error: Failure while executing; `\{\"SHELL\"=\>\"zsh\"\} /opt/homebrew/Cellar/adr-tools/3.0.0/bin/adr completion zsh` exited with 1. Here's the output:

```

**Comment by SMillerDev** [comment, marked as answer]
Homebrew doesn't ship patches that haven't been upstreamed. If upstream will never accept your patch we could switch the formula to a fork if one is accepted by the community as the successor.

**Reply by snim2** [reply]
Many thanks for such a quick and clear answer, that's a very sensible rule. I'll have a think about forks, unfortunately there are over 500 of them already!

**Reply by SMillerDev** [reply]
https://repology.org/ is generally a good place to start to see what the community is considering the successor.
